### DISCUSSION TITLE
Loading data into a Tree using async

### DISCUSSION THREAD

**Discussion by lolbat** [Textualize/textual#2730] [Q&A]
This probably isn't the most efficient way to do this.

I am loading some data when an app mounts and then using it to fill a Tree. It all works but the app is blank until the processes I am calling finish. It doesn't look too good.

This is the code I am using

```python
    def on_mount(self) -> None:

        self.loadPipData()

    def loadPipData(self) -> None:
  
        # load the pip data and display it
        if venvData.getVenvStatus():
            packageData = self.loadPackageData(packageType.VENV)
            self.displayPackageData(packageData, packageType.VENV)

        packageData = self.loadPackageData(packageType.GLOBAL)
        self.displayPackageData(packageData, packageType.GLOBAL)

        self.query_one(LoadingIndicator).display = False

    def loadPackageData(self, pipType: packageType) -> List[str]:

        # call to the pip reader and get it to list the packages in pip
        if pipType == packageType.GLOBAL:
            return pipReader.loadPipData(venvData.getPythonPath())
        else:
            # check to see if there is an active venv
            if venvData.getVenvStatus():
                # there is one so ask for the data from it
                return pipReader.loadPipData(venvData.getVenvPath())

    def displayPackageData(self, packages: List[str], pipType: packageType) -> None:

        # check the package data and if it is not empty then add a node for it

        nodeName = "global packages"
        if pipType == packageType.VENV:
            nodeName = "venv packages"

        if len(packages):
            pipTree = self.query_one(Tree)
            currentNode = pipTree.root.add(nodeName, expand=True)

            for aPackage in packages:
                currentNode.add(aPackage, expand=False, allow_expand=False)
```

The `loadPipData()` function is in a separate class and it uses a subprocess call to get a list of packages from pip.

I have been fooling around with async but, as you can expect, I am running into some issues using it. Currently all I have tried is using async in the `loadPipData()` and `loadPackageData()` functions. 

I assume that the issue is that I am having to await the subprocess while it runs. Would it be better to use `create_subprocess_exec` and do all the waiting there or should it be in the app itself and not a class?

**Comment by davep** [comment]
You may want to check out DirectoryTree. It loads up a Tree using workers.

**Reply by lolbat** [reply]
I'll check that out. Thanks. I was already reading the Workers section of the docs.

**Reply by davep** [reply]
Did that help?

**Reply by lolbat** [reply]
I have been RealWorld busy for the last little while. In that time I rewrote the code to use the sh library instead of subprocess and I now have to rewrite the main class that reads in the pip data so that it can do it as an async process. That is where I am currently at.
